common/rpc/interceptor/health_check.go

```diff
@@ -0,0 +1,127 @@
+package interceptor
+
+import (
+	"context"
+	"time"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/aggregate"
+	"go.temporal.io/server/common/log"
+	"google.golang.org/grpc"
+)
+
+// HealthCheckInterceptor is a gRPC interceptor that records health metrics
+type HealthCheckInterceptor struct {
+	historyHealthSignalAggregator HealthSignalAggregator
+}
+
+// NewHealthCheckInterceptor creates a new health check interceptor
+func NewHealthCheckInterceptor(historyHealthSignalAggregator HealthSignalAggregator) *HealthCheckInterceptor {
+	return &HealthCheckInterceptor{
+		historyHealthSignalAggregator: historyHealthSignalAggregator,
+	}
+}
+
+// UnaryIntercept implements the gRPC unary interceptor interface
+func (h *HealthCheckInterceptor) UnaryIntercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	startTime := time.Now()
+	resp, err := handler(ctx, req)
+	elapsed := time.Since(startTime)
+
+	h.historyHealthSignalAggregator.Record(elapsed, err)
+
+	return resp, err
+}
+
+type (
+	// HealthSignalAggregator interface for aggregating health signals
+	HealthSignalAggregator interface {
+		Record(latency time.Duration, err error)
+		AverageLatency() float64
+		ErrorRatio() float64
+	}
+
+	// HealthSignalAggregatorImpl implements HealthSignalAggregator
+	HealthSignalAggregatorImpl struct {
```

Comment:
nit: healthSignalAggregatorImpl

Reply:
Need to keep exported so that fx has access to it in the history package https://github.com/temporalio/temporal/blob/lanie/history-rpc-health-aggregator/service/history/fx.go#L190

Reply:
lmk if you think fx will keep a pointer to the same aggregator impl if it's passed in as an interface -- fx is a bit of a black box to me

Reply:
changed this to be unexported along with the other impl -- it needed an interface, not a struct